File: frontend-server/src/app/services/http/united/calendar-data.service.ts
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { CalendarDataDTO } from "../../../model/dto/united/calendar-data-dto";
import { convertDateStringToMoment } from "../schedule.service";
import { tap } from "rxjs/operators";
import { HttpClient } from "@angular/common/http";
import { RestConfig } from "../../../rest.config";

@Injectable({
  providedIn: 'root'
})
export class CalendarDataService {

  constructor(private httpClient: HttpClient,
              private config: RestConfig) {}

  getByEnterpriseIdAndDepartmentIdAndDate(enterpriseId: number,
                                          departmentId: number,
                                          startDate: string,
                                          endDate: string): Observable<CalendarDataDTO> {
    return this.httpClient.get<CalendarDataDTO>(
      `${this.config.baseUrl}/enterprises/${enterpriseId}/departments/${departmentId}
      /calendar_data?from=${startDate}&to=${endDate}`
    ).pipe(tap(value => value.schedule.forEach(dto => {
      dto.mainCompositions.forEach(convertDateStringToMoment);
      dto.substitutionCompositions.forEach(convertDateStringToMoment);
    })));
  }

  getByEnterpriseIdAndDepartmentIdAndShiftIdsAndDate(enterpriseId: number,
                                                     departmentId: number,
                                                     shiftIds: number[],
                                                     startDate: string,
                                                     endDate: string): Observable<CalendarDataDTO> {
    return this.httpClient.get<CalendarDataDTO>(
      `${this.config.baseUrl}/enterprises/${enterpriseId}/departments/${departmentId}/shifts/${shiftIds}
      /calendar_data?from=${startDate}&to=${endDate}`
    ).pipe(tap(value => value.schedule.forEach(dto => {
      dto.mainCompositions.forEach(convertDateStringToMoment);
      dto.substitutionCompositions.forEach(convertDateStringToMoment);
    })));
  }

}

File: work-schedule-service/src/main/java/com/averude/uksatse/scheduler/microservice/workschedule/service/united/CalendarDataService.java
package com.averude.uksatse.scheduler.microservice.workschedule.service.united;

import com.averude.uksatse.scheduler.microservice.workschedule.model.CalendarDataDTO;

import java.time.LocalDate;
import java.util.List;

public interface CalendarDataService {
    CalendarDataDTO getByDepartmentId(Long enterpriseId,
                                      Long departmentId,
                                      LocalDate from,
                                      LocalDate to);

    CalendarDataDTO getByShiftIds(Long enterpriseId,
                                  Long departmentId,
                                  List<Long> shiftIds,
                                  LocalDate from,
                                  LocalDate to);
}
